src/harness/sas_test_harness.py

```diff
@@ -0,0 +1,364 @@
+        if record_url is not None and generated_file_name is not None:
+          self.__writeDumpFile(generated_file_name, dump_record)
+          fad_record = self.__createFadRecord(record_url, dump_record)
+          fad_record_list.append(fad_record)
+        else:
+          raise Exception("SASTestHarnessError:Unable to create record url")
+      fad_file_name = 'FAD.json'
+      full_activity_dump = self.__createFadObject(fad_record_list)
+      self.__writeDumpFile(fad_file_name, full_activity_dump)
+      logging.info("FAD Records generated successfully."
+                   "The files are placed under the path:%s", self.getDumpFilePath())
+
+class SasHttpServer(HTTPServer):
+  """ The class SasHttpServer overrides built-in HTTPServer and takes the
+  parameter base_path used to serve the files. This is needed to have SAS
+  test harnesses to server files from different directories.
+  """
+  def __init__(self, base_path, server_address, RequestHandlerClass):
+    self.base_path = base_path
+    HTTPServer.__init__(self, server_address, RequestHandlerClass)
+
+class SasTestHarnessServerHandler(BaseHTTPRequestHandler):
+  """SasTestHarnessServerHandler class is inherited with BaseHTTPRequestHandler
+  to serve HTTP Response.
+  """
+
+  def __checkDumpFileExist(self, url_id):
+    """Check the url_id received in GET request passed as argument to check corresponding
+    dump file exist in that path.
+    """
+    return os.path.exists(os.path.join(self.server.base_path, '{}.json'.format(url_id)))
+
+  def __getDumpFilePath(self, url_id):
+    """ Return the absolute path of FAD records that are going
+    to be generated.
+    """
+    return os.path.join(self.server.base_path, '{}.json'.format(url_id))
+
+  def do_GET(self):
+    """Handles Pull/GET Request and returns Path of the Request to callback Method."""
+
+    if "dump" in self.path:
+      self.send_response(200)
+      self.send_header('Content-type', 'application/json')
+      self.end_headers()
+      self.wfile.write(json.load(open(self.__getDumpFilePath('FAD'))))
+    else:
+      url_encoded_id = self.path.split('/')[-1]
+      if not self.__checkDumpFileExist(url_encoded_id):
+        self.send_error(404)
+        return
+      self.send_response(200)
+      self.send_header('Content-type', 'application/json')
+      self.end_headers()
+      self.wfile.write(json.load(open(self.__getDumpFilePath(url_encoded_id))))
+
+# Helper functions for FAD related tests.
+def generateCbsdReferenceId(fcc_id, serial_number):
+  """creates encoded cbsd_reference_id using sha1 with fcc_id and serial number of cbsd.
+  """
+  return str(fcc_id + '/' + str(hashlib.sha1(serial_number).hexdigest())).encode('utf-8')
+
+def generateCbsdRecords(cbsd_records, grant_records_list):
+  """ Generates the cbsdData Object by combining cbsdrecord and grantrecord based on SAS-SAS TS.
+
+  Args:
+     cbsd_records: list of cbsdData objects that includes one or more grants object.
```

Comment:
Probably you mean that this is a list of RegistrationRequests (and that below you are using GrantRequests). Please be more explicit about the object types by using the names in the WInnForum specifications whenever possible.

Reply:
updated   the comment accordingly